silence/silence.go

```diff
@@ -134,47 +136,47 @@ func (s *Silencer) Mutes(lset model.LabelSet) bool {
 		// markerVersion because the Query call might already return a
 		// newer version, which is not the version our old list of
 		// applicable silences is based on.
-		sils, _, err = s.silences.Query(
-			QIDs(ids...),
-			QState(types.SilenceStateActive),
+		allIDs := append(append(make([]string, 0, totalSilences), activeIDs...), pendingIDs...)
+		allSils, _, err = s.silences.Query(
+			QIDs(allIDs...),
+			QState(types.SilenceStateActive, types.SilenceStatePending),
 		)
 	} else {
 		// New silences have been added, do a full query.
-		sils, newVersion, err = s.silences.Query(
-			QState(types.SilenceStateActive),
+		allSils, newVersion, err = s.silences.Query(
+			QState(types.SilenceStateActive, types.SilenceStatePending),
 			QMatches(lset),
 		)
 	}
 	if err != nil {
 		level.Error(s.logger).Log("msg", "Querying silences failed, alerts might not get silenced correctly", "err", err)
 	}
-	if len(sils) == 0 {
-		s.marker.SetSilenced(fp, newVersion)
+	if len(allSils) == 0 {
+		// Easy case, neither active nor pending silences anymore.
+		s.marker.SetSilenced(fp, newVersion, nil, nil)
 		return false
 	}
-	idsChanged := len(sils) != len(ids)
-	if !idsChanged {
-		// Length is the same, but is the content the same?
-		for i, s := range sils {
-			if ids[i] != s.Id {
-				idsChanged = true
-				break
-			}
-		}
-	}
-	if idsChanged {
-		// Need to recreate ids.
-		ids = make([]string, len(sils))
-		for i, s := range sils {
-			ids[i] = s.Id
+	// In still possible that nothing has changed, but finding out is not
+	// much less effort than just recreating the IDs from the query
+	// result. So let's do it in any case. Note that we cannot reuse the
+	// current ID slices for concurrency reasons.
+	activeIDs, pendingIDs = nil, nil
+	for _, sil := range allSils {
+		switch getState(sil, s.silences.now()) {
+		case types.SilenceStatePending:
+			pendingIDs = append(pendingIDs, sil.Id)
+		case types.SilenceStateActive:
+			activeIDs = append(activeIDs, sil.Id)
+		default:
+			// Do nothing, silence has expired in the meantime.
 		}
-		sort.Strings(ids) // For comparability.
 	}
-	if idsChanged || newVersion != markerVersion {
-		// Update marker only if something changed.
-		s.marker.SetSilenced(fp, newVersion, ids...)
-	}
-	return true
+	sort.Strings(activeIDs)
+	sort.Strings(pendingIDs)
```

Comment:
we no longer do any comparison of these so I don't think these need to sorted.

Reply:
But I think we are exposing the activeIDs via the API, and we shouldn't change the behavior here. pendingIDs are not exposed (they didn't exist before), but perhaps someone wants to add those later, so I thought I better sort them, too.

Reply:
Note: this could go in a comment.